File: preinstall/sbr-config/sbr_config/validator.py
"""Validation logic: check current state against correct SBR requirements."""

import logging
from typing import List, Optional

from .constants import TABLE_NAME_PREFIX
from .models import InterfaceInfo, Route, Rule, SystemState, ValidationResult
from .sysctl import validate_sysctl

logger = logging.getLogger(__name__)


def validate(state: SystemState) -> List[ValidationResult]:
    """Validate the current system state for correct source-based routing.

    Checks each non-default, non-loopback interface for:
    - Routing table entry exists
    - Subnet route in custom table
    - Default route in custom table
    - IP rule from interface IP to custom table
    - No conflicting rules

    Also validates sysctl settings.

    Args:
        state: The current system state from detector.

    Returns:
        List of ValidationResult entries (one per check).
    """
    results = []

    # Find the default route interface
    default_iface = _find_default_interface(state)

    # Validate each non-default interface
    for iface in state.interfaces:
        if iface.is_loopback:
            continue
        if iface.is_default_route_interface:
            results.append(ValidationResult(
                interface_name=iface.name,
                check_name="default_route_interface",
                is_correct=True,
                current_value=f"Default route via {iface.gateway or 'unknown'}",
                expected_value="Not modified by SBR",
                fix_description="",
            ))
            continue

        if not iface.is_up:
            results.append(ValidationResult(
                interface_name=iface.name,
                check_name="interface_state",
                is_correct=True,
                current_value="DOWN (skipped)",
                expected_value="Interface is down, skipping SBR checks",
                fix_description="",
            ))
            continue

        # Run SBR checks for this interface (with or without gateway)
        table_name = f"{TABLE_NAME_PREFIX}{iface.name}"
        results.extend(_validate_interface_sbr(iface, table_name, state))

    # Validate that the default route in main table is intact
    results.extend(_validate_default_route(state, default_iface))

    # Validate sysctl settings
    non_default_ifaces = [
        i.name for i in state.interfaces
        if not i.is_loopback and not i.is_default_route_interface and i.is_up
    ]
    results.extend(validate_sysctl(state.sysctl_values, non_default_ifaces))

    return results


def _find_default_interface(state: SystemState) -> Optional[InterfaceInfo]:
    """Find the interface carrying the default route."""
    for iface in state.interfaces:
        if iface.is_default_route_interface:
            return iface
    return None


def _validate_interface_sbr(
    iface: InterfaceInfo,
    table_name: str,
    state: SystemState,
) -> List[ValidationResult]:
    """Validate SBR configuration for a single interface."""
    results = []

    # Check 1: Routing table entry exists
    table_exists = any(
        rt.name == table_name for rt in state.routing_tables
    )
    results.append(ValidationResult(
        interface_name=iface.name,
        check_name="routing_table_exists",
        is_correct=table_exists,
        current_value=f"Table '{table_name}' {'exists' if table_exists else 'missing'}",
        expected_value=f"Table '{table_name}' in /etc/iproute2/rt_tables",
        fix_description=(
            f"Interface {iface.name} ({iface.ip_address}) needs a dedicated routing "
            f"table named '{table_name}' so traffic from this interface can be routed "
            f"independently of the main routing table."
        ) if not table_exists else "",
    ))

    # Check 2: Subnet route in custom table
    table_routes = state.routes_by_table.get(table_name, [])
    has_subnet_route = any(
        r.destination == iface.subnet and r.device == iface.name
        for r in table_routes
    )
    results.append(ValidationResult(
        interface_name=iface.name,
        check_name="subnet_route_in_table",
        is_correct=has_subnet_route,
        current_value=(
            f"Subnet route for {iface.subnet} "
            f"{'found' if has_subnet_route else 'missing'} in table {table_name}"
        ),
        expected_value=f"{iface.subnet} dev {iface.name} src {iface.ip_address} table {table_name}",
        fix_description=(
            f"Table '{table_name}' needs a route to the local subnet {iface.subnet} "
            f"via {iface.name} so the interface can reach its gateway."
        ) if not has_subnet_route else "",
    ))

    # Check 3: Default route in custom table (only if gateway is known)
    if iface.gateway is not None:
        has_default_route = any(
            r.destination == "default" and r.gateway == iface.gateway and r.device == iface.name
            for r in table_routes
        )
        results.append(ValidationResult(
            interface_name=iface.name,
            check_name="default_route_in_table",
            is_correct=has_default_route,
            current_value=(
                f"Default route via {iface.gateway} dev {iface.name} "
                f"{'found' if has_default_route else 'missing'} in table {table_name}"
            ),
            expected_value=f"default via {iface.gateway} dev {iface.name} table {table_name}",
            fix_description=(
                f"Table '{table_name}' needs a default route via {iface.gateway} "
                f"so traffic originating from {iface.ip_address} destined for remote "
                f"networks exits through {iface.name}'s gateway."
            ) if not has_default_route else "",
        ))
    else:
        # No gateway -- no default route needed (non-routable / subnet-only)
        results.append(ValidationResult(
            interface_name=iface.name,
            check_name="default_route_in_table",
            is_correct=True,
            current_value="No gateway -- default route not applicable",
            expected_value="No default route (non-routable interface)",
            fix_description="",
        ))

    # Check 4: IP rule exists
    has_rule = any(
        r.selector_from in (iface.ip_address, f"{iface.ip_address}/32")
        and r.table == table_name
        for r in state.rules
    )
    results.append(ValidationResult(
        interface_name=iface.name,
        check_name="ip_rule_exists",
        is_correct=has_rule,
        current_value=(
            f"Rule from {iface.ip_address} lookup {table_name} "
            f"{'found' if has_rule else 'missing'}"
        ),
        expected_value=f"from {iface.ip_address} lookup {table_name}",
        fix_description=(
            f"An IP rule is needed to direct traffic originating from "
            f"{iface.ip_address} to use table '{table_name}'. Without this rule, "
            f"the kernel uses the main routing table and responses may exit "
            f"via the wrong interface."
        ) if not has_rule else "",
    ))

    # Check 5: No conflicting rules
    conflicting = [
        r for r in state.rules
        if r.selector_from in (iface.ip_address, f"{iface.ip_address}/32")
        and r.table is not None
        and r.table != table_name
        and r.table not in ("main", "local", "default")
    ]
    has_conflict = len(conflicting) > 0
    if has_conflict:
        conflict_details = ", ".join(
            f"from {r.selector_from} lookup {r.table} (priority {r.priority})"
            for r in conflicting
        )
        results.append(ValidationResult(
            interface_name=iface.name,
            check_name="no_conflicting_rules",
            is_correct=False,
            current_value=f"Conflicting rules: {conflict_details}",
            expected_value="No rules from this IP pointing to other tables",
            fix_description=(
                f"Existing rules direct traffic from {iface.ip_address} to "
                f"table(s) other than '{table_name}'. This may conflict with "
                f"SBR configuration. Review and remove conflicting rules or "
                f"use --exclude {iface.name} to skip this interface."
            ),
        ))
    else:
        results.append(ValidationResult(
            interface_name=iface.name,
            check_name="no_conflicting_rules",
            is_correct=True,
            current_value="No conflicting rules found",
            expected_value="No conflicting rules",
            fix_description="",
        ))

    return results


def _validate_default_route(
    state: SystemState,
    default_iface: Optional[InterfaceInfo],
) -> List[ValidationResult]:
    """Validate that the main table's default route is intact."""
    results = []

    has_default = any(
        r.destination == "default" for r in state.routes_main
    )
    results.append(ValidationResult(
        interface_name="(main table)",
        check_name="default_route_intact",
        is_correct=has_default,
        current_value=(
            f"Default route {'present' if has_default else 'MISSING'} in main table"
        ),
        expected_value="Default route present in main table",
        fix_description=(
            "The main routing table has no default route. This means the system "
            "has no general internet connectivity path. SBR configuration should "
            "not remove the main table's default route."
        ) if not has_default else "",
    ))

    return results

File: preinstall/sbr-config/internal/validator/validator.go
// Package validator checks current state against correct SBR requirements.
package validator

import (
	"fmt"
	"log"
	"strings"

	"github.com/weka/tools/preinstall/sbr-config/internal/constants"
	"github.com/weka/tools/preinstall/sbr-config/internal/models"
	"github.com/weka/tools/preinstall/sbr-config/internal/sysctl"
)

// Validate validates the current system state for correct source-based routing.
func Validate(state *models.SystemState) []models.ValidationResult {
	var results []models.ValidationResult

	// Find the default route interface.
	defaultIface := findDefaultInterface(state)
	_ = defaultIface // used in default-route validation below

	// Validate each interface.
	for _, iface := range state.Interfaces {
		if iface.IsLoopback {
			continue
		}
		if iface.IsDefaultRouteInterface {
			gw := "unknown"
			if iface.Gateway != nil {
				gw = *iface.Gateway
			}
			results = append(results, models.ValidationResult{
				InterfaceName:  iface.Name,
				CheckName:      "default_route_interface",
				IsCorrect:      true,
				CurrentValue:   fmt.Sprintf("Default route via %s", gw),
				ExpectedValue:  "Not modified by SBR",
				FixDescription: "",
			})
			continue
		}

		if !iface.IsUp {
			results = append(results, models.ValidationResult{
				InterfaceName:  iface.Name,
				CheckName:      "interface_state",
				IsCorrect:      true,
				CurrentValue:   "DOWN (skipped)",
				ExpectedValue:  "Interface is down, skipping SBR checks",
				FixDescription: "",
			})
			continue
		}

		// Run SBR checks for this interface.
		tableName := constants.TableNamePrefix + iface.Name
		results = append(results, validateInterfaceSBR(&iface, tableName, state)...)
	}

	// Validate that the main table's default route is intact.
	results = append(results, validateDefaultRoute(state)...)

	// Validate sysctl settings.
	var nonDefaultIfaces []string
	for _, i := range state.Interfaces {
		if !i.IsLoopback && !i.IsDefaultRouteInterface && i.IsUp {
			nonDefaultIfaces = append(nonDefaultIfaces, i.Name)
		}
	}
	results = append(results, sysctl.ValidateSysctl(state.SysctlValues, nonDefaultIfaces)...)

	return results
}

func findDefaultInterface(state *models.SystemState) *models.InterfaceInfo {
	for i := range state.Interfaces {
		if state.Interfaces[i].IsDefaultRouteInterface {
			return &state.Interfaces[i]
		}
	}
	return nil
}

func validateInterfaceSBR(
	iface *models.InterfaceInfo,
	tableName string,
	state *models.SystemState,
) []models.ValidationResult {
	var results []models.ValidationResult

	// Check 1: Routing table entry exists.
	tableExists := false
	for _, rt := range state.RoutingTables {
		if rt.Name == tableName {
			tableExists = true
			break
		}
	}
	results = append(results, models.ValidationResult{
		InterfaceName: iface.Name,
		CheckName:     "routing_table_exists",
		IsCorrect:     tableExists,
		CurrentValue: fmt.Sprintf("Table '%s' %s", tableName,
			boolStr(tableExists, "exists", "missing")),
		ExpectedValue: fmt.Sprintf("Table '%s' in /etc/iproute2/rt_tables", tableName),
		FixDescription: func() string {
			if !tableExists {
				return fmt.Sprintf(
					"Interface %s (%s) needs a dedicated routing table named '%s' "+
						"so traffic from this interface can be routed independently "+
						"of the main routing table.",
					iface.Name, iface.IPAddress, tableName)
			}
			return ""
		}(),
	})

	// Check 2: Subnet route in custom table.
	tableRoutes := state.RoutesByTable[tableName]
	hasSubnetRoute := false
	for _, r := range tableRoutes {
		if r.Destination == iface.Subnet && r.Device == iface.Name {
			hasSubnetRoute = true
			break
		}
	}
	results = append(results, models.ValidationResult{
		InterfaceName: iface.Name,
		CheckName:     "subnet_route_in_table",
		IsCorrect:     hasSubnetRoute,
		CurrentValue: fmt.Sprintf("Subnet route for %s %s in table %s",
			iface.Subnet, boolStr(hasSubnetRoute, "found", "missing"), tableName),
		ExpectedValue: fmt.Sprintf("%s dev %s src %s table %s",
			iface.Subnet, iface.Name, iface.IPAddress, tableName),
		FixDescription: func() string {
			if !hasSubnetRoute {
				return fmt.Sprintf(
					"Table '%s' needs a route to the local subnet %s via %s "+
						"so the interface can reach its gateway.",
					tableName, iface.Subnet, iface.Name)
			}
			return ""
		}(),
	})

	// Check 3: Default route in custom table (only if gateway is known).
	if iface.Gateway != nil {
		gw := *iface.Gateway
		hasDefaultRoute := false
		for _, r := range tableRoutes {
			if r.Destination == "default" && r.Gateway != nil && *r.Gateway == gw && r.Device == iface.Name {
				hasDefaultRoute = true
				break
			}
		}
		results = append(results, models.ValidationResult{
			InterfaceName: iface.Name,
			CheckName:     "default_route_in_table",
			IsCorrect:     hasDefaultRoute,
			CurrentValue: fmt.Sprintf("Default route via %s dev %s %s in table %s",
				gw, iface.Name, boolStr(hasDefaultRoute, "found", "missing"), tableName),
			ExpectedValue: fmt.Sprintf("default via %s dev %s table %s",
				gw, iface.Name, tableName),
			FixDescription: func() string {
				if !hasDefaultRoute {
					return fmt.Sprintf(
						"Table '%s' needs a default route via %s so traffic "+
							"originating from %s destined for remote networks exits "+
							"through %s's gateway.",
						tableName, gw, iface.IPAddress, iface.Name)
				}
				return ""
			}(),
		})
	} else {
		results = append(results, models.ValidationResult{
			InterfaceName:  iface.Name,
			CheckName:      "default_route_in_table",
			IsCorrect:      true,
			CurrentValue:   "No gateway -- default route not applicable",
			ExpectedValue:  "No default route (non-routable interface)",
			FixDescription: "",
		})
	}

	// Check 4: IP rule exists.
	hasRule := false
	for _, r := range state.Rules {
		if r.SelectorFrom != nil && r.Table != nil &&
			(*r.SelectorFrom == iface.IPAddress || *r.SelectorFrom == iface.IPAddress+"/32") &&
			*r.Table == tableName {
			hasRule = true
			break
		}
	}
	results = append(results, models.ValidationResult{
		InterfaceName: iface.Name,
		CheckName:     "ip_rule_exists",
		IsCorrect:     hasRule,
		CurrentValue: fmt.Sprintf("Rule from %s lookup %s %s",
			iface.IPAddress, tableName, boolStr(hasRule, "found", "missing")),
		ExpectedValue: fmt.Sprintf("from %s lookup %s", iface.IPAddress, tableName),
		FixDescription: func() string {
			if !hasRule {
				return fmt.Sprintf(
					"An IP rule is needed to direct traffic originating from "+
						"%s to use table '%s'. Without this rule, the kernel uses "+
						"the main routing table and responses may exit via the wrong interface.",
					iface.IPAddress, tableName)
			}
			return ""
		}(),
	})

	// Check 5: No conflicting rules.
	var conflicting []models.Rule
	for _, r := range state.Rules {
		if r.SelectorFrom == nil || r.Table == nil {
			continue
		}
		from := *r.SelectorFrom
		tbl := *r.Table
		if (from == iface.IPAddress || from == iface.IPAddress+"/32") &&
			tbl != tableName &&
			tbl != "main" && tbl != "local" && tbl != "default" {
			conflicting = append(conflicting, r)
		}
	}

	if len(conflicting) > 0 {
		var details []string
		for _, r := range conflicting {
			details = append(details, fmt.Sprintf("from %s lookup %s (priority %d)",
				ptrStr(r.SelectorFrom), ptrStr(r.Table), r.Priority))
		}
		results = append(results, models.ValidationResult{
			InterfaceName: iface.Name,
			CheckName:     "no_conflicting_rules",
			IsCorrect:     false,
			CurrentValue:  fmt.Sprintf("Conflicting rules: %s", strings.Join(details, ", ")),
			ExpectedValue: "No rules from this IP pointing to other tables",
			FixDescription: fmt.Sprintf(
				"Existing rules direct traffic from %s to table(s) other than '%s'. "+
					"This may conflict with SBR configuration. Review and remove "+
					"conflicting rules or use --exclude %s to skip this interface.",
				iface.IPAddress, tableName, iface.Name),
		})
	} else {
		results = append(results, models.ValidationResult{
			InterfaceName:  iface.Name,
			CheckName:      "no_conflicting_rules",
			IsCorrect:      true,
			CurrentValue:   "No conflicting rules found",
			ExpectedValue:  "No conflicting rules",
			FixDescription: "",
		})
	}

	return results
}

func validateDefaultRoute(state *models.SystemState) []models.ValidationResult {
	hasDefault := false
	for _, r := range state.RoutesMain {
		if r.Destination == "default" {
			hasDefault = true
			break
		}
	}

	fix := ""
	if !hasDefault {
		fix = "The main routing table has no default route. This means the system " +
			"has no general internet connectivity path. SBR configuration should " +
			"not remove the main table's default route."
	}

	return []models.ValidationResult{{
		InterfaceName: "(main table)",
		CheckName:     "default_route_intact",
		IsCorrect:     hasDefault,
		CurrentValue: fmt.Sprintf("Default route %s in main table",
			boolStr(hasDefault, "present", "MISSING")),
		ExpectedValue:  "Default route present in main table",
		FixDescription: fix,
	}}
}

// Helpers

func boolStr(b bool, ifTrue, ifFalse string) string {
	if b {
		return ifTrue
	}
	return ifFalse
}

func ptrStr(s *string) string {
	if s != nil {
		return *s
	}
	return "<nil>"
}

func init() {
	// Suppress unused import warning for log.
	_ = log.Println
}
